### PR TITLE
#33 Copy Offload - Example C++ Client

### LINKED ISSUE #33: Copy Offload - Example C++ Client
Long-term we're going to drop the thin C layer we've put over gRPC Go.  This is not a tenable solution.
 
Instead, you should use the generated gRPC C++ client code.  This offers the following benefits:
1. Actual gRPC support, documentation, and plentiful examples can be found online.
2. Naming convention - The generated classes *.pb.{h,cc} classes are all scoped to the `datamovement` namespace, avoiding name collisions
3. Versioning - gRPC itself is pretty dynamic and referencing the *pb.{h,cc} directly would mean less reliance on maintaining the CGo library and resolve the versioning issue you raised.
   - Non-breaking changes, those that introduce a new fields, should be trivial.
   - Breaking changes would define a new method/message , i.e. CreateV2(), CreateRequestV2{},
4. Native asynchronous support - useful for Status requests with a non-zero timeout value.
 
While you can use the CGo library for early development and rapid prototyping, gRPC C++ should be used long term.

### PR DESCRIPTION
**Pull request by NateRoiger**
Closes #33 
Signed-off-by: Nate Roiger <nate.roiger@hpe.com>

### CODE REVIEW

File: daemons/compute/_client-c/Readme.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+**WARNING: The C client should not be used beyond rapid prototyping; The C++ client should be use long-term.**
+
+## Generating client code
+To generate the gRPC client interfaces from our .proto service definition, use the `protoc` protocol buffer compiler
+```
+$ make protoc
+```
+Running this command generate the following files in your current directory
+- datamovement.pb-c.h, the header which declares your generated message classes
+- datamovement.pb-c.c, which contains the implementation of your message classes
```
Comment (bdevcich-hpe) [inline daemons/compute/_client-c/Readme.md:14]:
```suggestion
- `datamovement.pb-c.h`, the header which declares your generated message classes
- `datamovement.pb-c.c`, which contains the implementation of your message classes
```

File: daemons/compute/client-cpp/Readme.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+```
+$ make datamovement.grpc.pb.o
+```
+
+Running this command generate the following files in your current directory
+- datamovement.pb.h, the header which declares your generated message classes
+- datamovement.pb.cc, which contains the implementation of your message classes
+- datamovement.grpc.pb.h, the header which declares your generated service classes
+- datamovement.grpc.pb.cc, which contains the implementation of your service classes
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/Readme.md:27]:
```suggestion
Running this command generates the following files in your current directory
- `datamovement.pb.h`, the header which declares your generated message classes
- `datamovement.pb.cc`, which contains the implementation of your message classes
- `datamovement.grpc.pb.h`, the header which declares your generated service classes
- `datamovement.grpc.pb.cc`, which contains the implementation of your service classes
```

File: daemons/compute/client-cpp/Readme.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+Collectively, these contain:
+- All the protocol buffer code to populate, serialize, and retrieve requests and response message types.
+- A class called DataMover that contains
+    - A remote interface type (or _stub_) for clients to call with the methods defined in the DataMover service
+    - Two abstract interfaces for servers to implement, also with the methods defined in the DataMover service
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/Readme.md:33]:
```suggestion
- A class called `DataMover` that contains
    - A remote interface type (or _stub_) for clients to call with the methods defined in the `DataMover` service
    - Two abstract interfaces for servers to implement, also with the methods defined in the `DataMover` service
```

File: daemons/compute/client-cpp/Readme.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+### Creating a stub
+To call service methods, we first need to create a stub.
+
+First we need to create a gRPC _channel_ for our stub, specifying the server address we want to connect to - in our case using a local socket with no SSL.
+
+```c++
+grpc::CreateChannel("unix:///tmp/nnf.sock", grpc::InsecureChannelCredentials()));
+```
+
+Now we can use the channel to create our stub using the NewStub method provided in the DataMover class we generated from our .proto.
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/Readme.md:45]:
```suggestion
Now we can use the channel to create our stub using the `NewStub` method provided in the `DataMover` class we generated from our .proto.
```

File: daemons/compute/client-cpp/Readme.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        std::unique_ptr<DataMover::Stub> stub_;
+};
+```
+
+### Calling service methods
+In this example we call the blocking/synchronous versions of each method: this means that the RPC calls wait for the server to respond.
+
+We call the following service methods:
+- Create Request / Response - Create an offload request; system implementors must ensure all parameters provided to the create request.
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/Readme.md:60]:
```suggestion
- Create Request / Response - Create an offload request; system implementors must ensure all parameters are provided to the create request.
```

File: daemons/compute/client-cpp/Readme.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+- Status Request / Response - Request status of an offload request.
+- Delete Request / Response - Request deletion of a completed offload request.
+
+
+## Compile and Run
+Build the client
+```
+$ make
+```
+Assuming you have a server running (i.e. `./server --simulated --socket=/tmp/nnf.sock`) Run the client
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/Readme.md:70]:
```suggestion
Assuming you have a server running (i.e. `./server --simulated --socket=/tmp/nnf.sock`), run the client
```

File: daemons/compute/client-cpp/client.cc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+        std::unique_ptr<DataMover::Stub> stub_;
+};
+
+int main(int argc, char** argv) {
+
+    DataMoverClient client(
+        grpc::CreateChannel("unix:///tmp/nnf.sock", grpc::InsecureChannelCredentials()));
+    
+
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/client.cc:55]: extra line

File: daemons/compute/client-cpp/client.cc [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+using datamovement::DataMover;
+using datamovement::DataMovementCreateRequest;
+using datamovement::DataMovementCreateResponse;
+using datamovement::DataMovementCreateResponse_Status;
+using datamovement::DataMovementStatusRequest;
+using datamovement::DataMovementStatusResponse;
+using datamovement::DataMovementStatusResponse_Status;
+using datamovement::DataMovementDeleteRequest;
+using datamovement::DataMovementDeleteResponse;
+using datamovement::DataMovementDeleteResponse_Status;
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/client.cc:40]: Delete and Status could be switched around to alphabetize

File: daemons/compute/client-cpp/datamovement.grpc.pb.cc [hunk 9]
```diff
@@ -0,0 +1,212 @@
+// Generated by the gRPC C++ plugin.
```
Comment (bdevcich-hpe) [inline daemons/compute/client-cpp/datamovement.grpc.pb.cc:1]: Should these 4 generated files be checked in? They can be generated and we've given instruction on how to generate them.

File: daemons/compute/server/main.go [hunk 10]
```diff
@@ -32,7 +32,7 @@ import (
 	"github.com/takama/daemon"
 	"google.golang.org/grpc"
 
-	pb "github.com/NearNodeFlash/nnf-dm/daemons/compute/api"
+	pb "github.com/NearNodeFlash/nnf-dm/daemons/compute/client-go/api"
```
Comment (bdevcich-hpe) [inline daemons/compute/server/main.go:35]: I like that you've removed the duplication.

**Comment by bdevcich-hpe** [comment]
Looks good. Mostly nitpicks on my part.

**Review by bdevcich-hpe — APPROVED** [review]

File: daemons/compute/client-cpp/datamovement.grpc.pb.cc (hunk 9 above)
Comment (NateRoiger) [reply daemons/compute/client-cpp/datamovement.grpc.pb.cc:1]: You're right they should be removed.

File: daemons/compute/client-cpp/client.cc (hunk 8 above)
Comment (NateRoiger) [reply daemons/compute/client-cpp/client.cc:40]: I prefer "order of use"

**Review by roehrich-hpe — APPROVED** [review]

**Pull request merged**